design/deploy-steps.md

```diff
@@ -0,0 +1,158 @@
+
+## Status
+
+implementable
+
+## Summary
+
+This design proposes customizing the deployment procedure by providing a high
+level abstraction around Ironic [deploy steps][deploy-steps], as well as
+allowing per-host deployment (IPA) images.
+
+[deploy-steps]: https://docs.openstack.org/ironic/latest/admin/node-deployment.html#deploy-steps
+
+## Motivation
+
+By default the Ironic agent downloads a provided image via HTTP(s), converts it
+(if needed) and writes directly to the disk. This procedure works for many
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer (`coreos-installer`).
+
+### Goals
+
+- A way to replace the default deploy procedure with a custom action.
+- A way to provide a link to the Ironic agent image per host.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Add a way to execute custom actions during deployment.
+- Building custom agent images.
+
+## Proposal
+
+### User Story
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New structure field `AgentImage` with the only field `URL`. It will accept
+  an optional link to a custom agent ISO image.
+- New structure field `CustomDeploy` with the only field `Method` to engage
+  a non-default deployment procedure as described below.
+- The `Image.URL` field will become optional if `CustomDeploy` is set
+  since we don't know if the deploy method actually requires a separate image.
+
+Setting `CustomDeploy` to a non-empty value will trigger deployment even if
+`Image.URL` is empty.
+
+### Implementation Details: AgentImage
+
+We will accept a custom agent ISO image to be used for each host in the new
+`AgentImage.URL` field (custom kernel/initramfs are out of scope for this
+proposal).
+
+When an image is updated for a deployed instance, a rebuild will be triggered
+(the same way it's done for `Image.URL`). When an image is updated in the
+*Ready* state, the node will be powered off, so that it can pick up the new
+image on the next deployment or cleaning attempt.
+
+To simplify the proposal we won't handle the situation when the agent image is
+updated in the middle of inspection or cleaning.
+
+### Implementation Details: CustomDeploy
+
+When `CustomDeploy` is set, we will use the new [custom-agent deploy
+interface][custom-agent] so that the default deploy steps do not apply. Then we
+request a deploy step with the name provided in the `CustomDeploy.Method`
+field. This step has to be provided in the agent image and is responsible for
+the whole deployment process.
+
+NOTE: `CustomDeploy` is a single-field structure to allow potential extensions
+in the future.
+
+Any required customization is expected to be provided via `UserData` and
+`NetworkData` or be built into the image itself.
```

Comment:
We should specify here how we are going to pass that data (and any other arguments) to the deploy step so that someone writing a deploy step knows what to expect.

Reply:
Ok